Integration tests

hardbyte opened the pull request:
We currently have *end to end* tests that use just the REST api, and a few (not many) *unit tests* that test isolated Python functions. 

This PR starts to add *integration tests* which have direct access to the support services - e.g. direct access to redis or postgres to check and/or manipulate the state. This is useful as it allows us to have smaller units of testable code relying on support services.

wilko77 commented on backend/entityservice/cache/progress.py:
```diff
@@ -1,13 +1,15 @@
 
 
-def save_current_progress(comparisons, run_id):
+def save_current_progress(comparisons, run_id, config=None):
```
I find it confusing to be able to pass `config` in here, as almost all of it gets ignored anyway. It'd be clearer if it's just a cache expiry value. But then, why would you want to set a different expiry value in the first place?

hardbyte replied:
Yeah you'd only want to do that for testing. My logic is that it would be nice to slowly move away from using the global `Config` as we add tests.

hardbyte commented on backend/entityservice/integrationtests/dbtests/test_insertions.py:
```diff
@@ -0,0 +1,93 @@
+
+    def _create_project(self):
+        project_id = generate_code()
+        project_auth_token = generate_code()
```
I'm embarrassed to say I forgot about those wonderful helpful models. I'll use the `Project` one here.

hardbyte commented on backend/entityservice/integrationtests/dbtests/test_insertions.py:
```diff
@@ -0,0 +1,93 @@
+        project = get_project(conn, project_id)
+        assert 'time_added' in project
+        assert not project['marked_for_deletion']
+        assert not project['uses_blocking']
```
Fair

hardbyte commented on backend/entityservice/integrationtests/dbtests/test_insertions.py:
```diff
@@ -0,0 +1,93 @@
+
+        end_time = time.perf_counter()
+        elapsed_time = end_time - start_time
+        assert elapsed_time < 1
```
I'll add a comment and increase it. On my desktop it takes <0.5s to insert 10k 128B elements + blocks into a local db. We both probably agree this may not be fast enough, but I wanted to start failing a test if it got any worse.

hardbyte commented on backend/entityservice/integrationtests/dbtests/test_insertions.py:
```diff
@@ -0,0 +1,93 @@
+        assert elapsed_time < 1
+
+        stored_encodings = list(get_encodingblock_ids(conn, dp_id, '1'))
+        assert len(stored_encodings) == num_entities
```
I actually changed the db a bit and added new selection functions that would be helpful in the branch feature-use-encodings-from-db so this will change. I'll update it to be clearer for now though - it is actually just pulling the encoding ids - not the encodings themselves back from the db.

hardbyte commented on backend/entityservice/integrationtests/dbtests/test_insertions.py:
```diff
@@ -0,0 +1,93 @@
+
+    def test_insert_project_and_dp(self):
+        project_id, project_auth_token, dp_id, dp_auth_token = self._create_project_and_dp()
+        assert len(dp_auth_token) == 48
```
fair - I'll remove it with the refactor to use the `Project` model anyway.

hardbyte commented:
Thanks for the review @wilko77 I think I've addressed your points well enough to merge. But happy to keep the conversation going as the integration tests take shape
